[World Leaks] Fix leaks in LayoutTests/imported/w3c/web-platform-tests/scroll-animations/

nipunshukla21 opened the pull request:
#### 56f36c24abfdf7d5b1be8dc65698c77064fe34a6
<pre>
[World Leaks] Fix leaks in LayoutTests/imported/w3c/web-platform-tests/scroll-animations/
<a href="https://bugs.webkit.org/show_bug.cgi?id=300333">https://bugs.webkit.org/show_bug.cgi?id=300333</a>
<a href="https://rdar.apple.com/162132950">rdar://162132950</a>

Reviewed by Antoine Quint.

Change WebAnimation::stop() such that upon stopping a WebAnimation we also remove the effect
by calling setEffectInternal(nullptr). This fixes approximately 10 leaky tests.

Test: animations/animation-timeline-does-not-leak.html

* LayoutTests/animations/animation-timeline-does-not-leak-expected.txt: Added.
* LayoutTests/animations/animation-timeline-does-not-leak.html: Added.
* LayoutTests/animations/resources/animation-timeline-leak-test.html: Added.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::stop):

Canonical link: <a href="https://commits.webkit.org/301956@main">https://commits.webkit.org/301956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a16556ea7a2c8c018a5d2f656d27bb03cec268af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126058 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132881 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77872 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46381 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95999 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64099 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129006 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112730 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76488 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36011 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30915 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76353 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106888 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31135 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135581 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52815 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104493 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108948 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104210 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26850 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49611 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27937 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/50193 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52712 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58545 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52046 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55393 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1640,6 +1640,7 @@ void WebAnimation::stop()
 {
     ActiveDOMObject::stop();
     removeAllEventListeners();
+    setEffectInternal(nullptr);
```
I'd wager we may need a check here to see whether the animation is currently considered [relevant](https://drafts.csswg.org/web-animations-1/#relevant-animations-section). This can be checked with `WebAnimation::m_isRelevant`. The scenario I'm thinking about is that an animation is started using the JavaScript API but no reference is ever kept to it, for instance by simply calling [`Element.animate()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/animate). I expect `WebAnimation::stop()` would be called here at some point and nuking the associated effect would have the visible effect or stopping an animation that should otherwise be alive.

nipunshukla21 replied:
Assuming I understood you correctly that an animation started by the JS API but without keeping a reference **is not relevant** it looks like adding an `if (!m_isRelevant)` block makes it leak again. Maybe there's an issue with how that state is set?

graouts replied:
You can ignore my comment:
```
bool WebAnimation::virtualHasPendingActivity() const
{
    // Keep the JS wrapper alive if the animation is considered relevant or could become relevant again by virtue of having a timeline.
    return m_timeline || m_isRelevant;
}
```
This makes sure that `WebAnimation::stop()` would _not_ be called for relevant animations.

nipunshukla21 committed:
[World Leaks] Fix leaks in LayoutTests/imported/w3c/web-platform-tests/scroll-animations/
https://bugs.webkit.org/show_bug.cgi?id=300333
rdar://162132950

Reviewed by Antoine Quint.

Change WebAnimation::stop() such that upon stopping a WebAnimation we also remove the effect
by calling setEffectInternal(nullptr). This fixes approximately 10 leaky tests.

Test: animations/animation-timeline-does-not-leak.html

* LayoutTests/animations/animation-timeline-does-not-leak-expected.txt: Added.
* LayoutTests/animations/animation-timeline-does-not-leak.html: Added.
* LayoutTests/animations/resources/animation-timeline-leak-test.html: Added.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::stop):

Canonical link: https://commits.webkit.org/301956@main
